---
title: "Lab 07"
author: "Biology Student"
date: "20 Feb 2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading Libraries

```{r Load Libraries, include=FALSE}
if (!require("tidyverse")) install.packages("tidyverse"); library(tidyverse)
if (!require("devtools")) install.packages("devtools"); library(devtools)
```

## Objectives for Lab 6

1. Introduction to RNA-Seq
2. Using Galaxy on BisonNet
3. Quality control of fastq files

## Background

We will be using the dataset the following paper: 
https://genome.cshlp.org/content/21/2/193.full

The paper is also available in the Readings directory of the class repo.

The Dataset is from Deep Sequencing of Poly(A)+ RNA from the Drosophila melanogaster S2-DRSC cells that have been RNAi depleted of mRNAs encoding RNA binding proteins. 

For the tutorial, you will need use the following files in Galaxy: 
Untreated: SRR031711, SRR031714, SRR031716
RNAi: SRR031718, SRR031724, SRR031726

These files were obtained from the NCBI's Sequence Read Archive: https://www.ncbi.nlm.nih.gov/sra
Look up each of the Run Accession numbers and record the needed information below.

The number of reads is listed as "# of Spots".

```{r Summary}
SRAsummary <- data.frame(matrix(ncol = 5, nrow = 0))
names(SRAsummary) <- c("SampleName", "LongName", "Layout", "Reads", "Bases")
SRAsummary %>% 
  add_row(
    SampleName = "SRR031711", 
    LongName = "S2_DRSC_Untreated-1", 
    Layout = "SINGLE", 
    Reads = 6064911, 
    Bases = 272.9 * 10^6) -> SRAsummary
```

Calculate how long each read is expected to be:

```{r}
SRAsummary %>%
  mutate(ReadLength = round(Bases / (Reads * (1 + (Layout == "PAIRED"))))) -> SRAsummary
print(SRAsummary)
write_csv(SRAsummary, path = "SRAsummary.csv")
```

## Study Design

Review the material at https://rnaseq.uoregon.edu/, in the paper, and in the SRA to answer the following questions:

1. What are the conditions that we are going to compare? 

2. How many biological replicates are we studying for each condition?

3. What sequencing platform was used?

4. What selection scheme (if any) was performed on the RNA?

5. What was the depth of sequencing for each sample (min, max, and median)?

6. Was a stranded library prep protocol used? If so, what is the strandedness of the reads?

7. Were the sequences single-end or paired end?

8. What is the length of the reads?

## Galaxy!

Log in at http://galaxy.bucknell.edu

You must be on campus or connected to the VPN to access Galaxy on BisonNet.

Take the interactive tours on the `Galaxy UI` and the `History`.

## Moving shared data into your history

Click on the `Shared Data > Data Libraries` tab.

Click the Pasilla Datset. Select all of the files and then `To History > as Datasets`.

Click `Analyze Data` and verify that the 10 data files have been moved to the History.

Click the eye icon next to each to verify that the files look like they are in the proper fastq format and that the reads are the expected length:  
https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html

## Running FastQC

In the Tools section, type FastQC in the search box and then click the matching tool.

Run FastQC on one of the files and then wait for it to complete.

View the results by clicking on the eye icon for the Webpage output.

Run FastQC on each of the fastq files.

When all of the jobs are complete (this may take a little while), run the MultiQC tool to aggregate all of the FastQC results into a single report. 
To select multiple files as input, use command-click (Mac) or ctrl-click (PC).

After MultiQC is run compare the samples to determine if there are any large differences in quality problems between them. 
You may need to download the MultiQC webpage, unzip it, and open the html file in a browser. 
Download a file from Galaxy by clicking on it in the History area and then clicking the disk icon.

## QC Summary

http://www.bioinformatics.babraham.ac.uk/projects/fastqc/

Review the examples of "Good" and "Bad" Illumina data and summarize your impressions of this dataset. 

## Trimmomatic

http://www.usadellab.org/cms/index.php?page=trimmomatic

To try to improve the quality of this dataset, we will use trimmomatic. 
Trimmomatic performs a variety of useful trimming tasks for illumina paired-end and single ended data. 

Use Galaxy to run trimmomatic on each of the files, paying close attention to the PAIRED files and processing them as `two separate input files`. 

Use the following trimmomatic parameters (http://journal.frontiersin.org/article/10.3389/fgene.2014.00013/full):

1. "Perform initial ILLUMINACLIP step?" should be changed to Yes and the settings changed to TruSeq2 (single or paired, as appropriate) and 2:40:15:8. Note that the adapter sequences to be used should match the sequencing technology used to obtain the reads. According to the SRA metadata, these files were obtained using a Genome Analyzer II (GAII) and so TruSeq2 is the correct primer file.

2. "SLIDINGWINDOW" should be changed to "Average quality required" = 5

3. Then "Insert trimmomatic operation" and set it to "MINLEN" = 25

Run trimmomatic on each set of files (a total of 6 times).

For the paired files, trimmomatic will produce 4 output files: R1 Paired, R2 Paired, R1 Unpaired, and R2 Unpaired. 
The unpaired files contain reads that were orphaned when their mate was discarded due to the trimming settings. 
They can be used in some analyses, but we will not be needing them, so it is best to delet them at this point so they don't confuse us later.

## FastQC and MultiQC

Now run FastQC and MultiQC on the trimmed data. Has it improved? Is it "good" yet?

## Helpful RNA-Seq Links

RNA-seqlopedia: https://rnaseq.uoregon.edu/

RNA-Seq Blog: http://www.rna-seqblog.com/

QCFAIL Blog: https://sequencing.qcfail.com/
(Unfortunately it looks like they are no longer posting, but they have some great posts about problems with certain Illumina sequencers.)

QCFAIL post about SRA file corruption: https://sequencing.qcfail.com/articles/data-can-be-corrupted-upon-extraction-from-sra-files/


