---
title: "Lab 13"
author: "Biology Student"
date: "15 Apr 2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading Libraries

```{r Load Libraries, include=FALSE}
if (!require("tidyverse")) install.packages("tidyverse"); library(tidyverse)
if (!require("devtools")) install.packages("devtools"); library(devtools)
```

## Objectives for Lab 13

1. Introduction to Gene Ontology
2. Generate a gene set
3. Conduct Gene Set Enrichment Analysis using g:Profiler
4. Visualize a gene network using Cytoscape

## Background

Read the overview of Gene Ontology: 
http://geneontology.org/docs/ontology-documentation/

Determine whether you are interested in examining just one, or all three of the gene ontology domains.

## Gene Sets

The results from a differential expression analysis is a list of genes (or transcripts) that meet a given set of criteria (such as FDR cutoff and minimum fold-change). 
We are often interested in whether there is a pattern among the functions of these genes - whether entire pathways have been up- or down- regulated. 

From the Pasilla analysis, we identified a number of genes that were differentially expressed after RNAi treatment. 
From the WNS bat study, we identified a number of genes that were differentially expressed in tissue infected with the WNS fungus compared to uninfected tissue.

```{r DE Results}
warmPd_posvswarmPd_neg_complete <- read_delim("../04-Data/warmPd-posvswarmPd-neg.complete.txt", "\t", escape_double = FALSE, trim_ws = TRUE)
warmPd_posvswarmPd_neg_complete %>%
  filter(padj <= 0.05 & FC >= 2) %>%
  arrange(-FC) %>%
  select(Id) -> upregulated
write_csv(upregulated, path="../04-Data/warmPd_posvswarmPd_neg_upregulated.csv")
```

## g:Profiler

https://biit.cs.ut.ee/gprofiler/gost

g:GOSt performs functional enrichment analysis, also known as over-representation analysis (ORA) or gene set enrichment analysis, on input gene list. It maps genes to known functional information sources and detects statistically significantly enriched terms.

1. Launch g:GOSt and use the csv file to paste in the list of genes of interest.
2. Change the Organism to *Myotis lucifugus*
3. Select "Ordered query"
4. Click "Data sources" and select the appropriate gene ontology domains. This is also where you can download a gmt file to be used for network analysis in the next section.
5. Click Run query

The output will indicate categories that meet the adjusted p value cutoff selected under "Advanced settings". 
Mouse over each category to view its name. 
Clicking "Detailed results" will list each enriched category and the genes associated with it. 

If the number of categories is large (>50) you may wish to narrow the list by decreasing the "Term size" in the "Detailed results" section. Lower the upper threshold until fewer than 50 results are displayed.

Download the results as both a csv and a gem file. 

## Network visualization

https://cytoscape.org

Based on https://enrichmentmap.readthedocs.io/en/docs-2.2/Tutorial_GProfiler.html

Download Cytoscape to your computer and follow the installation instructions. You may have to upgrade or install Java. After it is installed, launch Cytoscape.

1. In the Apps menu, select App Manager and then install EnrichmentMap
2. In the menu bar, locate the App tab and then select > EnrichmentMap > Create Enrichment Map
3. Make sure the Analysis Type is set to "generic" (ex:gProfiler)
4. Please select the following files by clicking on the respective (â€¦) button and selecting the file in the Dialog:
- GMT / The gmt file you downloaded
- Dataset 1 / Enrichments: The gem file you downloaded
- Tune Parameters
- P-value cut-off: 1
- Q-value cut-off: 1
- Overlap Coefficient cut-off: 0.5
5. Click on the Build radio button at the bottom of the panel to create the Enrichment Map
6. In the menu bar, Go to View, and activate Show Graphics Details
7. In the control panel, go to Style, click on Label and select EM1_GS_DESCR in the Column dropdown. This will label nodes with names rather than GO IDs.
